spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/amqp/RabbitStreamConnectionDetails.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2012-2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.boot.autoconfigure.amqp;
+
+import org.springframework.boot.autoconfigure.service.connection.ConnectionDetails;
+
+/**
+ * Details required to establish a connection to a RabbitMQ service.
```

Comment:
Why do we need this? Especially given the javadoc, it looks like to me that the existing `RabbitConnectionDetails` should be enough, isn't it?

Reply:
rabbitmq stream uses a different port and in order to do that the plugin should be enabled. Having two implementations of `RabbitConnectionDetails` will lead to having an exception with 2 beans unless other beans or configuration are disabled when stream are enabled.

I think would be possible to have a different virtual host for stream in a single RabbitMQ instance, so, having a specific `ConnectionDetails` would make sense.

Reply:
I think there's a bug in the current code as we're not taking `RabbitConnectionDetails` into account when falling back from stream-specific configuration to general Rabbit configuration. I've opened https://github.com/spring-projects/spring-boot/issues/42489 for that. When fixed, we'll fallback from `spring.rabbit.stream.*` properties to `RabbitConnectionDetails`. With the changes here in place, we'd then fallback from `RabbitStreamConnectionDetails` (a replacement for some of the `spring.rabbit.stream.*` properties) to `RabbitConnectionDetails` (a replacement for some of the general `spring.rabbit.*` properties).

Reply:
Thinking about this some more, Rabbit Streams support is deceptively complicated (to me at least) as I think it's the first time where different users may want different things. One user may want `RabbitConnectionDetails` and `RabbitStreamConnectionDetails` to be sourced from the same container as their app is using the same Rabbit broker for both streaming and regular messaging. Another user may want `RabbitConnectionDetails` and `RabbitStreamConnectionDetails` to be sourced from different containers as their app is using different Rabbit brokers for streaming and regular messaging.

With only `RabbitMQContainer` to look at, even if we could detect in the container connection details factory that the streams plugin is enabled, we wouldn't know if the user wanted this container to be used only for streams or for both streams and regular messaging. It feels like we may need some other signal to indicate how the container should be used as a source for connection details. This reminds me a bit of the [exploratory work being done for SSL](https://github.com/spring-projects/spring-boot/issues/41137) and the possibility of having SSL-related annotations on the container field. I'll flag this one for a meeting so I can see what the rest of the team thinks.

Reply:
We think we may be able to check to see if the default streams port is exposed by the container. If it isn't, the factory should return `null`.

Reply:
Upon further thought, I'm not sure that this will work for all cases. Those cases are:

1. One container to be used for both streams and "standard" messaging
2. Two containers: one for streams and one for standard messaging

In the two container case, `RabbitContainerConnectionDetailsFactory` will match a container with the streams port exposed and I think we'll end up with a duplicate connection details failure. If we made `RabbitContainerConnectionDetailsFactory` ignore containers with the streams port exposed it would fix the two container case but it would break the one container case.

Some experimentation's needed to see if the above's accurate.